### PR TITLE
Fixes hair sticking through hats.

### PR DESCRIPTION
**Pull request by kingofkosmos**
Fixes hair sticking through hats.

Before:
capture
After:
capture2

This is done by using the dynamic_hair flag. Dynamic_hair set to nothing wherever the hair needs to stick through (kitty ears etc).

:cl:
fix: Fixed hair sticking through headgear.
/:cl:

### CODE REVIEW

**Comment by kingofkosmos** [comment]
I just accidentally did "Merge branch 'master' into hiddenhairunderhats", hope it's ok?

File: code/modules/clothing/head/collectable.dm [hunk 1]
```diff
@@ -4,6 +4,7 @@
 /obj/item/clothing/head/collectable
 	name = "collectable hat"
 	desc = "A rare collectable hat."
+	dynamic_hair_suffix = "+detective"
```
Comment (ChangelingRain) [inline code/modules/clothing/head/collectable.dm]: this is inherited.

File: code/modules/clothing/head/beanie.dm [hunk 2]
```diff
@@ -6,6 +6,7 @@
 	desc = "A stylish beanie. The perfect winter accessory for those with a keen fashion sense, and those who just can't handle a cold breeze on their heads."
 	icon_state = "beanie" //Default white
 	item_color = "beanie"
+	dynamic_hair_suffix = "+detective"
```
Comment (ChangelingRain) [inline code/modules/clothing/head/beanie.dm]: also inherited.

File: code/modules/clothing/head/collectable.dm (hunk 1 above)
Comment (kingofkosmos) [inline code/modules/clothing/head/collectable.dm]: Really? From where?

File: code/modules/clothing/head/jobs.dm [hunk 3]
```diff
@@ -1,3 +1,5 @@
+/obj/item/clothing/head/
```
Comment (ChangelingRain) [inline code/modules/clothing/head/jobs.dm]: really, from where?

Comment (kingofkosmos) [inline code/modules/clothing/head/jobs.dm]: No need to be sassy, I was genuinely asking a question. Is that ok there or should it be moved to here?
 https://github.com/tgstation/tgstation/blob/eace4eec148e8a366ee2fe6673f852b59a2e77b1/code/modules/clothing/clothing.dm#L187

Comment (lzimann) [inline code/modules/clothing/head/jobs.dm]: Move this variable to the class definition [here](https://github.com/tgstation/tgstation/blob/2da8867f67a669509b40055a404047877afb5b00/code/modules/clothing/clothing.dm#L269)

**Comment by kingofkosmos** [comment]
All commented things fixed and tested through.

File: code/modules/clothing/head/collectable.dm [hunk 4]
```diff
@@ -14,6 +14,7 @@
 	name = "collectable slime cap!"
 	desc = "It just latches right in place!"
 	icon_state = "slime"
+	dynamic_hair_suffix = "null"
```
Comment (RemieRichards) [inline code/modules/clothing/head/collectable.dm]: Don't do this, do ```dynamic_hair_suffix = ""```
At the moment you're loading the image called "null", which doesn't exist, <i>so it just happens to 'work', as in, not do anything as you'd expect</i> however in situations like this you should use true null ```null```, but since it's a string, ```""``` instead, so that the dynamic hair code knows to instead fallback to the object's HIDE_HAIR flag (if applicable)

tl;dr ```dynamic_hair_suffix  = "null"``` -> ```dynamic_hair_suffix = ""```

**Comment by RemieRichards** [comment]
Not a bad idea to use the detective sprites for all hats.
though I'd prefer they (all) be renamed to +generic (even for the detective) since they're now serving that purpose.

**Comment by kingofkosmos** [comment]
>Not a bad idea to use the detective sprites for all hats.
though I'd prefer they (all) be renamed to +generic (even for the detective) since they're now serving that purpose.

Thank you! This overlay was very easy to work with!

Only problem I encountered that some "shallow" hats like the beret or chef's hat look a bit wonky with certain hairs because the hair stops one pixel too early, which looks like the person's balding hair is peeking out a bit from below the hat. But it's very minor, I'll post a pic of it later today.

I'll rename all the "+detective"'s to "+generic".

**Comment by RemieRichards** [comment]
@kingofkosmos in those cases, I would say not to use dynamic hair, just let them render the full hair (which I think is what berets and things do, where as they used to hide all hair which looked UGLY AS SHIT)

We can fix situations like that with specific hair extensions (like how +det was originally just for the det hat)

**Comment by kingofkosmos** [comment]
No beret:
capture
Beret with dynamic_hair:
capture2
No dynamic_hair:
capture3
You're right, it does look better with full hair. I'll change the beret and check which others looked wonky.

**Comment by WJohn** [comment]
So this fixes #3583 super dead finally? Awesome!

**Comment by kingofkosmos** [comment]
Oh nice this actually bring the issue number down! Added that to the main post.

**Comment by lzimann** [comment]
That issue is already closed.

**Comment by kingofkosmos** [comment]
oh ok I didn't know

**Comment by DoctoreMad** [comment]
Does this remove Det's hat awful "300 NEW DMIs FOR ONE HAT" solution?

**Comment by kingofkosmos** [comment]
If you mean dynamic hair, this doesn't overwrite it, only applies it to most hats.

**Comment by lzimann** [comment]
At least now you use those 91~ sprites for more than one thing, as opposed to a single item(which is really awful). So it's an improvement I guess.
